src/agent/coverage/src/debuginfo.rs

```diff
@@ -103,20 +103,72 @@ impl ModuleDebugInfo {
             }
         }
 
-        // Now we're sure we want this data, so leak it.
+        // Now we're more sure we want this data. Leak it so the parsed object
+        // will have a `static` lifetime.
         let data = Box::leak(data);
 
-        let object = Object::parse(data)?;
+        // Save a raw pointer to the file data. If object parsing fails, or
+        // there is no debuginfo, we will use this to avoid leaking memory.
+        let data_ptr = data as *mut _;
+
+        let object = match Object::parse(data) {
+            Ok(object) => {
+                if !object.has_debug_info() {
+                    // Drop the object to drop its static references to the leaked data.
+                    drop(object);
```

Comment:
If we fail to do this before rehydrating/dropping the `Box`, it won't necessarily cause observable memory corruption on drop, but _is_ undefined behavior, since the internal references will be dangling.